src/utils/cudnn.cpp

```diff
@@ -328,23 +184,26 @@ void TensorDescriptor::create() {
 
 void TensorDescriptor::set(
   cudnnDataType_t data_type,
-  const std::vector<int>& dims,
-  std::vector<int> strides) {
+  std::vector<int> dims_in,
+  std::vector<int> strides_in) {
 
   // Check that arguments are valid
-  if (dims.empty()) {
+  if (dims_in.empty()) {
     LBANN_ERROR("attempted to set cuDNN tensor descriptor with no dimensions");
   }
-  if (dims.size() < 3) {
-    // As of cuDNN 7.65, cuDNN does not support tensors with <3 dims
-    LBANN_ERROR(
-      "attempted to set cuDNN tensor descriptor with fewer than 3 dimensions");
-  }
-  if (!strides.empty() && dims.size() != strides.size()) {
+  if (!strides_in.empty() && dims_in.size() != strides_in.size()) {
     LBANN_ERROR(
       "attempted to set cuDNN tensor descriptor ",
-      "with mismatched dimensions (",dims.size(),") ",
-      "and strides (",strides.size(),")");
+      "with mismatched dimensions (", dims_in.size(), ") ",
+      "and strides (",strides_in.size(),")");
+  }
+
+  std::vector<int> dims=std::move(dims_in), strides=std::move(strides_in);
+  if (dims.size() < 3)
+  {
+    dims.resize(3, 1);
+    if (strides.size() < 3)
+      strides.resize(3, 1);
```

Comment:
It would be somewhat better to prepend the ones.

Reply:
The issue is that the first dim is the size, so it's "inter-pending". Suppose `dims_in = {7,5}`. Do you have 7 samples with 5 1x1 feature maps (this would match `develop`, or do you have 7 samples with 1 1x5 (or 5x1?) feature maps? As I mentioned, `develop` will extend this to `{7,5,1,1}`, so I was trying to match that behavior. (At least I got rid of one of the extraneous 1s, so it's just `{7,5,1}` after this PR!)

Reply:
Hm, matching existing behavior in `develop` is sensible. I'm also thinking about passing the data as-is into cuDNN and letting it throw an error, since it's really an internal problem in cuDNN's implementation.

Reply:
Unless we have a way to open a ticket with them and complain, I'm not in favor of throwing the error. This affects models as simple as Lenet, that we have to be able to run. At one level or another, this is something we need to deal with. Since it's a fiddly bit that's specific to cuDNN, I think it belongs here, far away from the actual layer implementation, which ought not concern itself with cuDNN at all.